Replace double buffering in OSD with single buffer+bitmask

fiam opened the pull request:
Needs testing, since there are significant timing changes in the OSD code.

fiam committed:
Use do {} while style macros

Rename macros ENABLE_MAX7456 and DISABLE_MAX7456 to function-like
macros and wrap their definitions in do {} while(0) blocks.

Should result in the same binary, but code is cleaner this way.

fiam committed:
Use a bitarray for screen updates in MAX7456

Avoids double buferring the screen. Saves around 456 bytes
of BSS in exchange for ~50 bytes of text.

fiam committed:
Fix potential bug in OSD character flushing

Should make screen way more responsive, since there were lot of
updates which didn't flush any character.

Code was checking that the amount of (characters checked+characters
actually sent to MAX7456) was < MAX_CHARS2UPDATE.
Note that this could be something tuned to make each update to the OSD
shorter less time in exchange for more updates required to send a given
number of characters. However, due to spiBuff being MAX_CHARS2UPDATE*6
and each character update being 6 bytes, it strongly suggests that
this was actually a bug because spiBuff would be MAX_CHARS2UPDATE*3
otherwise (3 read+3 written).

digitalentity commented on src/main/drivers/max7456.c:
```diff
@@ -155,24 +156,48 @@
+// to uint32_t.
+STATIC_ASSERT(VIDEO_BUFFER_CHARS_PAL % 32 == 0, VIDEO_BUFFER_CHARS_PAL_not_multiple_of_32);
+
+#define BYTE_ARRAY_SET_BY_WORDS(arr, x) \
```
`memset` of newlib-nano is smart enough to use 32-bit access for fastest possible operation. The only downside of using `memset` is a function call overhead.

fiam replied:
I hadn't bothered checking the source of `memset`. I assumed the code setting the buffer by aliasing to a 32bit ptr was there for some good reason, but now that you mention it I think it might be a leftover from the MAX7456 handling in the Arduino-based projects. Using `memset` saves 32 bits of flash, so let's change it.

digitalentity replied:
@fiam Since both `screenBuffer` and `screenIsDirty` are `uint8_t` arrays and it would be good to add an `__attribute__ ((aligned (4)))` to their definition to make sure `memset` won't have to deal with mis-aligned bytes.

digitalentity commented on src/main/drivers/max7456.c:
```diff
@@ -407,15 +430,28 @@ uint8_t* max7456GetScreenBuffer(void) {
+        //do not write over screen
+        if (x+i < 30) {
+            c = *buff;
+            pos = y*30+x+i;
```
`y*30+x` can be pre-calculated outside the loop

fiam replied:
Nice catch! We can also just increment `pos` once per iteration rather than always adding `i` and break out of the loop once `x+i >= 30`. Code should be cleaner this way too.

digitalentity commented on src/main/drivers/max7456.c:
```diff
@@ -479,15 +515,18 @@ void max7456DrawScreenPartial(void)
+        // Note that k should not be incremented on each
+        // iteration, only when we find a character that
+        // we need to send to the OSD.
+        for (k=0; k< MAX_CHARS2UPDATE;) {
```
This needs to be profiled to ensure `max7456DrawScreenPartial` doesn't block for too long

fiam replied:
I'll leave that up to you, since I don't have an STLink available ATM.

digitalentity replied:
Will do

fiam replied:
Do we have a macro for alignment? I can't find it, but I might be looking in the wrong places (or  grepping for the wrong strings, should I say).

If we don't, I'd like to introduce one in `common/utils.h`, since writing bare attributes looks a bit  dirty to me.

digitalentity replied:
No, we don't have a macro for alignment

digitalentity commented on src/main/drivers/max7456.c:
```diff
@@ -502,10 +541,10 @@ void max7456DrawScreenPartial(void)
-            ENABLE_MAX7456;
+            ENABLE_MAX7456();
             for (k=0; k < buff_len; k++)
                 spiTransferByte(MAX7456_SPI_INSTANCE, spiBuff[k]);
```
`spiTransfer` should be more efficient for bulk transfers

digitalentity replied:
We can also optimize spiTransfer to use 16-bit transfer where possible, but that's a scope of another PR

fiam replied:
It seems DMA is enabled depending on the target configuration. Only `SPRACINGF3NEO` seems to use it (both for TX and RX). Do all boards which might define `MAX7456_DMA_CHANNEL_TX` also allow sending the MAX7456 commands via SPI? Also note that we'll be sending 30 bytes at most (36 if we end up using HW blinking).

digitalentity replied:
In many cases MAX shares the bus with gyro which makes using DMA tricky at best. As long as the code does check for `dmaTransactionInProgress` we can mix DMA'ed and polled access.

digitalentity reviewed (changes requested):
Looks good in general. Please comment on suggested changes.

fiam committed:
Add macro for __attribute__ ((aligned(x)))

fiam committed:
Align screenBuffer and screenIsDirty to 4 bytes

Should allow memset() to work faster on them

fiam committed:
Reduce MAX_CHARS2UPDATE to 5

After fixing the bug on the previous commit, updates were taking
too long because they could actually end up sending MAX_CHARS2UPDATE
characters over SPI, which takes too long. This value should still
be responsive the aircraft while flying while performing fast
updates. Eventually, we'll introduce different update pauses
for ARMED/DISARMED states, which should make CMS menus faster.

Kudos to @digitalentity for the profiling

fiam committed:
Transfer bytes to the MAX7456 via SPI faster

Use spiTransfer() rather than spiTransferByte().

As requested by @digitalentity

fiam committed:
Remove outdated comment

Buffers are filled via plain-old memset
